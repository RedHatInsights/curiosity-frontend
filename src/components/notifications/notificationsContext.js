import { useContext, useCallback } from 'react';
import { NotificationsContext } from '@redhat-cloud-services/frontend-components-notifications';
import { helpers } from '../../common';

/**
 * @memberof Notifications
 * @module NotificationsContext
 */

/**
 * Use platform notifications. Apply a convenience wrapper for easily removing notifications based on an internal
 * "swatchId"
 *
 * @param {NotificationsContext} [context=NotificationsContext]
 * @returns {{
 *     addNotification: Function,
 *     clearNotifications: Function,
 *     removeNotification: removeNotification }} Add, clear all, or remove a notification.
 *
 *     - `addNotification` - Add a toast notification. A `swatchId` property is exposed to allow for easy removal.
 *     - `clearNotifications` - Clear all notifications
 *     - `removeNotifications` - Remove a toast notification based on ID. If you used a plain text `swatchId` to add
 *         the notification, this can be used to remove it. If `DEV_MODE` is enabled, a warning is logged to the
 *         console when an attempt is made to remove a notification using an invalid or non-existent `swatchId` or `id`.
 */
const useNotifications = ({ context = NotificationsContext } = {}) => {
  const { addNotification, removeNotification, getNotifications, ...contextMethods } = useContext(context);

  return {
    ...contextMethods,
    /**
     * Add a toast notification.
     *
     * A `swatchId` property is exposed to allow for easy removal.
     *
     * @param {object} notification - Notification object to be added.
     * @param {string} [notification.swatchId] - Optional plain language "unique" identifier that allows for
     *     easy removal.
     * @param {string} [variant] - Optional variant to display, defaults to "info"
     * @param {React.ReactNode} title - Notification title
     * @param {React.ReactNode} [description] - Notification description
     * @returns {void}
     */
    addNotification: useCallback(
      notification => {
        const { swatchId, swatchid, ...remainingNotification } = notification;
        return addNotification({ ...remainingNotification, swatchid: swatchId || swatchid });
      },
      [addNotification]
    ),

    /**
     * Remove a toast notification.
     *
     * For convenience IF a `swatchId` property is provided of the notification.
     *
     * @param {id} notification - Unique identifier to remove. This can be the plain language swatchId, or
     *     the generatedId provided by the notification package.
     */
    removeNotification: useCallback(
      id => {
        const notifications = getNotifications();
        const notification = notifications.find(
          ({ swatchid: internalId, id: generatedId }) => internalId === id || generatedId === id
        );

        if (notification) {
          removeNotification(notification.id);
        } else if (helpers.DEV_MODE) {
          console.warn(
            `Notification with id "${id}" not found. Make sure the notification was created with the "swatchId" prop, or you used the generated id provided by notifications.`
          );
        }
      },
      [getNotifications, removeNotification]
    )
  };
};

const context = {
  useNotifications
};

export { context as default, context, useNotifications };
