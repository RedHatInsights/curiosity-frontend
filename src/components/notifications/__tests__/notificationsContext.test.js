import React from 'react';
import { context, useNotifications } from '../notificationsContext';

describe('NotificationsContext', () => {
  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should apply a hook wrapper for notifications', async () => {
    const { result } = await renderHook(() => useNotifications());
    expect(result).toMatchSnapshot('notifications hook');
  });

  it.each([
    {
      description: 'swatchId',
      notification: {
        swatchId: 'loremIpsum',
        title: 'Lorem ipsum'
      }
    },
    {
      description: 'swatchid',
      notification: {
        swatchid: 'dolorSitAmet',
        title: 'Lorem ipsum'
      }
    },
    {
      description: 'id',
      notification: {
        id: 'ametConsectetur',
        title: 'Lorem ipsum'
      }
    },
    {
      description: 'swatchId and existing notification',
      notification: {
        swatchId: 'ametConsectetur',
        title: 'Lorem ipsum'
      },
      notifications: [
        {
          id: 'generated-id-1',
          swatchid: 'ametConsectetur',
          title: 'The original notification'
        },
        {
          id: 'generated-id-2',
          swatchid: 'loremIpsum'
        },
        {
          id: 'generated-id-3',
          swatchid: 'dolorSit'
        }
      ]
    }
  ])('should attempt addNotification with a custom ID: $description', async ({ notification, notifications = [] }) => {
    const mockAddNotification = jest.fn();
    const mockRemoveNotification = jest.fn();
    const mockGetNotifications = jest.fn(() => [...notifications]);
    const mockClearNotifications = jest.fn();

    const mockUseContext = jest.spyOn(React, 'useContext').mockReturnValue({
      addNotification: mockAddNotification,
      removeNotification: mockRemoveNotification,
      getNotifications: mockGetNotifications,
      clearNotifications: mockClearNotifications
    });

    const { result } = await renderHook(() => useNotifications());
    result.addNotification({
      ...notification
    });

    expect({
      addNotifications: mockAddNotification.mock.calls,
      removeNotifications: mockRemoveNotification.mock.calls,
      getNotifications: mockGetNotifications.mock.results
    }).toMatchSnapshot();
    mockUseContext.mockClear();
  });

  it.each([
    {
      description: 'swatchId',
      findNotification: 'loremIpsum',
      notifications: [
        {
          id: 'generated-id-1',
          swatchid: 'loremIpsum',
          title: 'Lorem ipsum'
        }
      ]
    },
    {
      description: 'id',
      findNotification: 'generated-id-2',
      notifications: [
        {
          id: 'generated-id-2',
          swatchid: 'dolorSitAmet',
          title: 'Lorem ipsum'
        }
      ]
    },
    {
      description: 'incorrect id',
      findNotification: 'incorrectId',
      notifications: [
        {
          id: 'generated-id-3',
          swatchid: 'ametConsectetur',
          title: 'Lorem ipsum'
        }
      ]
    }
  ])(
    'should find a notification with getNotification and hasNotification: $description',
    async ({ notifications, findNotification }) => {
      const mockAddNotification = jest.fn();
      const mockRemoveNotification = jest.fn();
      const mockGetNotifications = jest.fn(() => [...notifications]);
      const mockClearNotifications = jest.fn();

      const mockUseContext = jest.spyOn(React, 'useContext').mockReturnValue({
        addNotification: mockAddNotification,
        removeNotification: mockRemoveNotification,
        getNotifications: mockGetNotifications,
        clearNotifications: mockClearNotifications
      });

      const { result } = await renderHook(() => useNotifications());
      expect({
        getNotification: result.getNotification(findNotification),
        hasNotification: result.hasNotification(findNotification)
      }).toMatchSnapshot();
      mockUseContext.mockClear();
    }
  );

  it.each([
    {
      description: 'swatchId',
      removeId: 'loremIpsum',
      notifications: [
        {
          id: 'generated-id-1',
          swatchid: 'loremIpsum',
          title: 'Lorem ipsum'
        }
      ]
    },
    {
      description: 'id',
      removeId: 'generated-id-2',
      notifications: [
        {
          id: 'generated-id-2',
          swatchid: 'dolorSitAmet',
          title: 'Lorem ipsum'
        }
      ]
    },
    {
      description: 'incorrect id',
      removeId: 'incorrectId',
      notifications: [
        {
          id: 'generated-id-3',
          swatchid: 'ametConsectetur',
          title: 'Lorem ipsum'
        }
      ]
    }
  ])('should attempt removeNotification with a custom ID: $description', async ({ notifications, removeId }) => {
    const mockAddNotification = jest.fn();
    const mockRemoveNotification = jest.fn();
    const mockGetNotifications = jest.fn(() => [...notifications]);
    const mockClearNotifications = jest.fn();

    const mockUseContext = jest.spyOn(React, 'useContext').mockReturnValue({
      addNotification: mockAddNotification,
      removeNotification: mockRemoveNotification,
      getNotifications: mockGetNotifications,
      clearNotifications: mockClearNotifications
    });

    const { result } = await renderHook(() => useNotifications());
    result.removeNotification(removeId);

    expect(mockRemoveNotification.mock.calls).toMatchSnapshot();
    mockUseContext.mockClear();
  });
});
