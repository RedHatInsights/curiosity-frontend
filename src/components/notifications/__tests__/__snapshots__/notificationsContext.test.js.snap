// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`NotificationsContext should apply a hook wrapper for notifications: notifications hook 1`] = `
{
  "addNotification": [Function],
  "clearNotifications": [Function],
  "getNotification": [Function],
  "hasNotification": [Function],
  "removeNotification": [Function],
  "subscribe": [Function],
}
`;

exports[`NotificationsContext should attempt addNotification with a custom ID: id 1`] = `
{
  "addNotifications": [
    [
      {
        "id": "ametConsectetur",
        "swatchId": undefined,
        "title": "Lorem ipsum",
      },
    ],
  ],
  "getNotifications": [
    {
      "type": "return",
      "value": [],
    },
  ],
  "removeNotifications": [],
}
`;

exports[`NotificationsContext should attempt addNotification with a custom ID: swatchId (legacy swatchid) 1`] = `
{
  "addNotifications": [
    [
      {
        "swatchId": undefined,
        "swatchid": "dolorSitAmet",
        "title": "Lorem ipsum",
      },
    ],
  ],
  "getNotifications": [
    {
      "type": "return",
      "value": [],
    },
  ],
  "removeNotifications": [],
}
`;

exports[`NotificationsContext should attempt addNotification with a custom ID: swatchId 1`] = `
{
  "addNotifications": [
    [
      {
        "swatchId": "loremIpsum",
        "title": "Lorem ipsum",
      },
    ],
  ],
  "getNotifications": [
    {
      "type": "return",
      "value": [],
    },
  ],
  "removeNotifications": [],
}
`;

exports[`NotificationsContext should attempt addNotification with a custom ID: swatchId and existing notification 1`] = `
{
  "addNotifications": [
    [
      {
        "swatchId": "ametConsectetur",
        "title": "Lorem ipsum",
      },
    ],
  ],
  "getNotifications": [
    {
      "type": "return",
      "value": [
        {
          "id": "generated-id-1",
          "swatchId": "ametConsectetur",
          "title": "The original notification",
        },
        {
          "id": "generated-id-2",
          "swatchId": "loremIpsum",
        },
        {
          "id": "generated-id-3",
          "swatchId": "dolorSit",
        },
      ],
    },
  ],
  "removeNotifications": [
    [
      "generated-id-1",
    ],
  ],
}
`;

exports[`NotificationsContext should attempt removeNotification with a custom ID: id 1`] = `
[
  [
    "generated-id-2",
  ],
]
`;

exports[`NotificationsContext should attempt removeNotification with a custom ID: incorrect id 1`] = `[]`;

exports[`NotificationsContext should attempt removeNotification with a custom ID: swatchId 1`] = `
[
  [
    "generated-id-1",
  ],
]
`;

exports[`NotificationsContext should find a notification with getNotification and hasNotification: id 1`] = `
{
  "getNotification": {
    "id": "generated-id-2",
    "swatchId": "dolorSitAmet",
    "title": "Lorem ipsum",
  },
  "hasNotification": true,
}
`;

exports[`NotificationsContext should find a notification with getNotification and hasNotification: incorrect id 1`] = `
{
  "getNotification": undefined,
  "hasNotification": false,
}
`;

exports[`NotificationsContext should find a notification with getNotification and hasNotification: swatchId 1`] = `
{
  "getNotification": {
    "id": "generated-id-1",
    "swatchId": "loremIpsum",
    "title": "Lorem ipsum",
  },
  "hasNotification": true,
}
`;

exports[`NotificationsContext should return specific properties: specific properties 1`] = `
{
  "useNotifications": [Function],
}
`;
